### ISSUE TITLE
potential memory leak issue in subscription 

### ISSUE THREAD

**Issue by lsping8** [apollographql/apollo-client#7775]
Hi, I have this test code to test out if Apollo cache is causing the memory leak issue. I have a subscription resolver that will keep publishing at the interval of 10ms .

```
import { useMutation, useQuery, useSubscription } from '@apollo/client'
import { QUERY_TEST, SUB_TEST, MUTATE_TEST } from './documentNodes'

export default function Home() {
  const { data, loading } = useQuery(QUERY_TEST);
  const [startTest] = useMutation(MUTATE_TEST)
  useSubscription(SUB_TEST)

  if (loading) return null
  return (
    <div>
      <div><button onClick={() => startTest()}>start test</button></div>
      {JSON.stringify(data, null, 2)}
    </div>
  )
}
```

**Intended outcome:**

The heap memory in use should not keep increasing and not dropping at all.

**Actual outcome:**

This screenshot is taken when the page is active for a few minutes, when it's started the heap size is around 30mb.
<a href="https://ibb.co/ssx0Zdx"><img src="https://i.ibb.co/8PvWHLv/Screenshot-2021-02-27-at-9-30-50-PM.png" alt="Screenshot-2021-02-27-at-9-30-50-PM" border="0" /></a>
**How to reproduce the issue:**

https://github.com/lsping8/apollo-memory-test
I have created this repo to reproduce the issue, clone the repo and follow the setup process. In the app I have added a button to start interval process, click on the `start test` button at the left top corner and monitor the memory in chrome.

**Versions**

  System:
    OS: Linux 4.15 Ubuntu 16.04.7 LTS (Xenial Xerus)
  Binaries:
    Node: 12.21.0 - /usr/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 6.14.11 - /usr/bin/npm
  Browsers:
    Firefox: 86.0
  npmPackages:
    @apollo/client: ^3.3.11 => 3.3.11 
    apollo-server: ^2.21.0 => 2.21.0 
    apollo-upload-client: ^14.1.3 => 14.1.3

**Comment by brainkim** [comment]
Hey @lsping8! Thank you so much for creating a reproduction, and wanted to tell you I’m looking into this. I tried running your example and I need to look a bit deeper into it, but it seems like Chrome or Apollo Client will collect the memory at some point, creating a saw-tooth pattern in terms of memory usage. So my question is, have you noticed the sample application or your actual code actually crashing from running out of memory? If you haven’t seen it do this, I’m hesitant to call it a leak, so much as maybe using too much memory. I, of course, could be wrong about this :smile:

Also note that if you want more consistent memory usage, the cache abstraction provides some utility methods to reclaim and evict objects as described in the following document https://www.apollographql.com/docs/react/caching/garbage-collection/

Again, thanks for opening this issue, and just wanted to let you know this issue is seen and heard.

**Comment by lsping8** [comment]
Hi @brainkim, thank you for the reply. To answer your question if the application is crashing, the answer is yes. Well not the sample repo, but the production application we have.

<a href="https://ibb.co/qMykxyv">
<img src="https://i.ibb.co/8DK6XKH/Whats-App-Image-2021-02-26-at-4-24-15-PM.jpg" alt="Whats-App-Image-2021-02-26-at-4-24-15-PM" border="0" />
</a>

Firefox seem to be less tolerance at this issue. As for workaround, we just set all subscription to `no-cache` fetchPolicy and it seem to resolve the issue.

**Comment by brainkim** [comment]
Potentially related:
https://github.com/apollographql/apollo-client/issues/7013

**Comment by fourjuaneight** [comment]
We're seeing similar behavior on our production app as well. We're running Angular with Universal on Heroku. And there's a peak at about 1.5GBs of memory and then it crashes. Tried several patches with various other packages that could've been the culprit, but from our latest findings, seems like Apollo is likely causing it.

**Comment by ncioj10** [comment]
@brainkim Hi, we're also experiencing this issue. Is there any news on that?

**Comment by israelidanny** [comment]
Yup, experiencing this issue HARD on react native

**Comment by phryneas** [comment]
The aforementioned change is now included in Apollo Client 3.9 beta  - please try it out and give us feedback! :)

You can read up more on that in our [blog post about the AC 3.9 memory story](https://www.apollographql.com/blog/apollo-3-9-beta-feature-spotlight-the-memory-story).

**Comment by phryneas** [comment]
3.9 has been released, so I will close this issue for now. 

If this keeps happening in AC => 3.9.0, please open a new issue :)
